cloud/cluster/cluster.go

```diff
@@ -5,121 +5,153 @@ package cluster
+	stat stats.StatsReceiver
 
-// Cluster's ch channel accepts []Node and []NodeUpdate types, which then
-// get passed to its state to either SetAndDiff or UpdateAndFilter
-func NewCluster(state []Node, updateCh chan ClusterUpdate) *Cluster {
-	s := makeState(state)
-	c := &Cluster{
-		state:    s,
-		reqCh:    make(chan interface{}),
-		updateCh: updateCh,
-		subs:     nil,
-	}
-	go c.loop()
-	return c
-}
+	updateFreq time.Duration
 
-func (c *Cluster) Members() []Node {
-	ch := make(chan []Node)
-	c.reqCh <- ch
-	return <-ch
-}
+	// latestFetchedNodesCh, contains the latest node list from fetcher.
+	fetchedNodesCh chan []Node
 
-func (c *Cluster) Subscribe() Subscription {
-	ch := make(chan Subscription)
-	c.reqCh <- ch
-	return <-ch
-}
+	// nodesReqCh, used (by groupcache) to request the current list of nodes
+	NodesReqCh NodeReqChType
+
+	// nodesUpdatesChan, the node updates that the scheduler needs to process.
+	NodesUpdatesCh chan []NodeUpdate
+
+	useNodesUpdatesCh bool
+	bufferSize        int
 
-func (c *Cluster) Close() error {
-	close(c.reqCh)
-	return nil
+	priorNodeUpdateTime  time.Time
+	priorFetchUpdateTime time.Time
 }
 
-func (c *Cluster) done() bool {
-	return c.reqCh == nil && c.updateCh == nil
+// NewCluster creates a Cluster object and starts its processing loop and returns either a nodes updates channel
+// (reporting node add/remove events) or nodes list channel (reporting the current list of nodes).
+// The processing loop continuously gets the latest nodes list from the fetched nodes channel and either
+// put the list of added/removed nodes on the nodes updates channel or if the nodes list is requested put the
+// current list of nodes on the nodes list channel
+func NewCluster(stat stats.StatsReceiver, fetcher Fetcher, useNodesUpdatesCh bool, updateFreq time.Duration, chBufferSize int) (chan []NodeUpdate, NodeReqChType) {
+	fetchedNodesCh := StartFetchCron(fetcher, updateFreq, chBufferSize, stat)
+
+	s := makeState([]Node{})
+	var updatesCh chan []NodeUpdate = nil
+	var nodesReqCh NodeReqChType = nil
+	if useNodesUpdatesCh {
+		updatesCh = make(chan []NodeUpdate, chBufferSize)
+	} else {
+		nodesReqCh = make(NodeReqChType)
+	}
+	c := &Cluster{
+		state:                s,
+		updateFreq:           updateFreq,
+		stat:                 stat,
+		fetchedNodesCh:       fetchedNodesCh,
+		NodesReqCh:           nodesReqCh,
+		NodesUpdatesCh:       updatesCh,
+		useNodesUpdatesCh:    useNodesUpdatesCh,
+		priorNodeUpdateTime:  time.Now(),
+		priorFetchUpdateTime: time.Now(),
+		bufferSize:           chBufferSize,
+	}
+	if stat == nil {
+		c.stat = stats.NilStatsReceiver()
+	}
+	// logging
+	chDesc := []string{}
+	if c.NodesUpdatesCh != nil {
+		chDesc = append(chDesc, "NodesUpdateCh")
+	}
+	if c.NodesReqCh != nil {
+		chDesc = append(chDesc, "NodesReqCh")
+	}
+	log.Infof("cluster loop starting with frequency %s, reporting on %s with channel buffer size %d", c.updateFreq, strings.Join(chDesc[:], ","), chBufferSize)
```

Comment:
nits: chDest doesn't need to be a slice. Also the "[:]" in the Join call is unnecessary.

Reply:
I fixed [:].  Using slice for chDest to get 'nice' commas printed in the log statement